packages/flutter/lib/src/painting/text_painter.dart

```diff
@@ -270,17 +270,163 @@ class _UntilTextBoundary extends TextBoundary {
   }
 }
 
+class _TextLayout {
+  _TextLayout._(this._paragraph);
+
+  // This field is not final because the owner TextPainter could create a new
+  // ui.Paragraph with the exact same text layout (for example, when only the
+  // color of the text is changed).
+  ui.Paragraph _paragraph;
+
+  // TODO(LongCatIsLooong): https://github.com/flutter/flutter/issues/31707
+  // remove this hack as well as the flooring in `layout`.
+  static double _applyFloatingPointHack(double layoutValue) => layoutValue.ceilToDouble();
+
+  /// Whether this layout has been invalidated and disposed.
+  ///
+  /// Only for use when asserts are enabled.
+  bool get debugDisposed => _paragraph.debugDisposed;
+
+  /// The horizontal space required to paint this text.
+  ///
+  /// If a line ends with trailing spaces, the trailing spaces may extend
+  /// outside of the horizontal paint bounds defined by [width].
+  double get width => _applyFloatingPointHack(_paragraph.width);
+
+  /// The vertical space required to paint this text.
+  double get height => _applyFloatingPointHack(_paragraph.height);
+
+  /// The width at which decreasing the width of the text would prevent it from
+  /// painting itself completely within its bounds.
+  double get minIntrinsicLineExtent => _applyFloatingPointHack(_paragraph.minIntrinsicWidth);
+
+  /// The width at which increasing the width of the text no longer decreases the height.
+  ///
+  /// Includes trailing spaces if any.
+  double get maxIntrinsicLineExtent => _applyFloatingPointHack(_paragraph.maxIntrinsicWidth);
+
+  /// The distance from the left edge of the leftmost glyph to the right edge of
+  /// the rightmost glyph in the paragraph.
+  double get longestLine => _applyFloatingPointHack(_paragraph.longestLine);
+
+  /// Returns the distance from the top of the text to the first baseline of the
+  /// given type.
+  double getDistanceToBaseline(TextBaseline baseline) {
+    switch (baseline) {
+      case TextBaseline.alphabetic:
+        return _paragraph.alphabeticBaseline;
+      case TextBaseline.ideographic:
+        return _paragraph.ideographicBaseline;
+    }
+  }
+}
+
+class _TextPainterLayoutCache {
+  _TextPainterLayoutCache(this.layout, this.inputWidth, this.textAlignment, double minWidth, double maxWidth, TextWidthBasis widthBasis)
+    : contentWidth = _contentWidthFor(minWidth, maxWidth, widthBasis, layout),
+      assert(textAlignment >= 0.0 && textAlignment <= 1.0);
+
+  final _TextLayout layout;
+  double inputWidth;
```

Comment:
It could be helpful to future readers to document some of the fields on this class. What's the inputWidth? What's the offset? etc